### ISSUE TITLE
regex for parsing images from html has some flaws

### ISSUE THREAD

**Issue by nosilver4u** [Automattic/jetpack#9947]
Hi, I was looking through the regex for parsing images from the HTML recently, and noticed a couple flaws.
For reference, here's the regex:
`#(?:<a[^>]+?href=["|\'](?P<link_url>[^\s]+?)["|\'][^>]*?>\s*)?(?P<img_tag><img[^>]*?\s+?src=["|\'](?P<img_url>[^\s]+?)["|\'].*?>){1}(?:\s*</a>)?#is`

Notably, the character classes for quotation marks (single/double) have pipe characters in them. Normally, a pipe character indicates "alternation", but a character class already implies alternation and interprets a pipe character as a literal character instead. This means it would match an attribute surrounded by pipe characters, like so:
`<a href=|https://www.example.com/image.jpg|>`

Of course, it isn't likely to ever encounter that pattern as it isn't valid HTML, but the | should be removed from all of the character classes.

Secondly, the HTML spec says that the equals sign for an attribute is followed or preceded by zero or more whitespace characters, so it would be good to account for those, even if they are not used very often.

The corrected regex would look like this:

`#(?:<a[^>]+?href\s*=\s*["\'](?P<link_url>[^\s]+?)["\'][^>]*?>\s*)?(?P<img_tag><img[^>]*?\s+?src\s*=\s*["\'](?P<img_url>[^\s]+?)["\'].*?>){1}(?:\s*</a>)?#is`

**Comment by ebinnion** [comment]
Thank you for the report as well as the extra detail @nosilver4u! 👍 

When I did a search for the regex, I found it in the Photon module, so I've added Photon as a label.

**Comment by nosilver4u** [comment]
Indeed, direct link that I forgot to include: https://github.com/Automattic/jetpack/blob/37c33d71653e6e2b2e3083b8228c53d20c59afb1/class.photon.php#L82
